src/components/street-segment.js

```diff
@@ -0,0 +1,424 @@
+    }
+  },
+  updateSurfaceFromType: function (typeObject) {
+    // update color, surface, level from segment type preset
+    this.el.setAttribute(
+      'street-segment',
+      `surface: ${typeObject.surface}; color: ${typeObject.color}; level: ${typeObject.level};`
+    ); // to do: this should be more elegant to check for undefined and set default values
+  },
+  updateGeneratedComponentsList: function () {
+    // get all components on entity with prefix 'street-generated'
+    let generatedComponentList = [];
+    const components = this.el.components;
+    for (const componentName in components) {
+      if (componentName.startsWith('street-generated')) {
+        generatedComponentList.push(componentName);
+      }
+    }
+    this.generatedComponents = generatedComponentList;
+  },
+  update: function (oldData) {
+    const data = this.data;
+    const dataDiff = AFRAME.utils.diff(oldData, data);
+    // if oldData is same as current data, then don't update
+    if (AFRAME.utils.deepEqual(oldData, data)) {
+      return;
+    }
+    // regenerate components if only type has changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('type')
+    ) {
+      let typeObject = this.types[this.data.type];
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      this.remove();
+      this.createGeneratedComponentsFromType(typeObject); // add components for this type
+      this.updateSurfaceFromType(typeObject); // update surface color, surface, level
+    }
+    // propagate change of direction to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('direction')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'direction', this.data.direction);
+      }
+    }
+    // propagate change of length to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('length')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'length', this.data.length);
+      }
+    }
+    this.clearMesh();
+    this.height = this.calculateHeight(data.level);
+    this.tempXPosition = this.el.getAttribute('position').x;
+    this.el.setAttribute('position', { x: this.tempXPosition, y: this.height });
+    this.generateMesh(data);
+    // if width was changed, trigger re-justification of all street-segments by the managed-street
+    if (Object.keys(dataDiff).includes('width')) {
+      this.el.parentNode.components['managed-street'].applyJustification();
```

Comment:
If you want to decouple the two components. You can emit a widthChanged event and listen to it in managed-street. Also manage-street component may not exist on the parent entity?

Reply:
I confused street-segment with previous street component. street-segment can't be created without managed-street.